fix: fix resourcePerNode override not applied with Volcano scheduler

sksingh2005 opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, check our contributor guidelines: https://www.kubeflow.org/docs/about/contributing
2. To know more about Kubeflow Trainer, check the developer guide:
    https://github.com/kubeflow/trainer/blob/master/CONTRIBUTING.md
3. If you want *faster* PR reviews, check how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:
- Computes PodGroup.spec.minResources for Volcano using TrainJob.spec.trainer.resourcesPerNode for the trainer PodSet, instead of the base TrainingRuntime template.
- Continues to derive minMember from spec.trainer.numNodes , preserving existing behavior.
- Adds a unit test to validate that minResources reflects per-job overrides (multiplied by the trainer node count).

**Why We Need It**
- Ensures gang scheduling reserves the correct resources for each job instance, avoiding under-reservation that can delay or mis-schedule distributed training.
- Aligns with the expected Volcano behavior where PodGroup.minResources equals the sum of pod resource requests for all members, matching users’ intent when they set resourcesPerNode at submission time.
- Improves consistency between runtime defaults and job-level overrides, bringing the computed PodGroup in line with the actual pods that JobSet will run.

**Which issue(s) this PR fixes** 
Fixes #2980 
Fixes https://github.com/kubeflow/trainer/issues/2525


**Checklist:**

- [x] [Docs](https://www.kubeflow.org/docs/components/trainer/) included if any changes are user facing

andreyvelich commented on pkg/runtime/framework/plugins/volcano/volcano.go:
```diff
@@ -157,13 +158,22 @@ func (v *Volcano) Build(ctx context.Context, info *runtime.Info, trainJob *train
+			} else if res.Limits != nil && len(res.Limits) > 0 {
+				perPod = res.Limits
+			}
+		}
```
Thanks for this work @sksingh2005!
I think, we should consider approach to update PodSet resources in the MLPolicy plugins as I shared here: https://github.com/kubeflow/trainer/issues/2980#issuecomment-3543370153

Ideally, `SinglePodRequests` should have correct value once we create PodGroup object for Volcano/Co-scheduling. We can also consider to rename `SinglePodRequests` to `SinglePodResources` to cover requests/limits.

We might also want to use this value further in the extension framework, when we create JobSet: https://github.com/kubeflow/trainer/blob/a364e318539bb5d49384a04f23a8938e0e91c41f/pkg/runtime/framework/plugins/jobset/builder.go#L144

cc @astefanutti @tenzen-y

sksingh2005 replied:
Updated the changes sir

sksingh2005 replied:
sir i did not understood "Also, we should consider to switch to SinglePodResources to cover requests and limits assignment."

sksingh2005 commented on pkg/runtime/runtime.go:
```diff
@@ -139,13 +139,13 @@ func WithPodSet(
+			Ancestor:           ancestor,
+			Count:              ptr.To(max(count, 1)),
+			Volumes:            podSpecApply.Volumes,
+			SinglePodResources: resourcehelpers.PodRequests(&corev1.Pod{Spec: typedPodSpec}, resourcehelpers.PodResourcesOptions{}),
```
@andreyvelich sir done the changes as you mentioned.

andreyvelich commented on pkg/runtime/framework/plugins/mpi/mpi.go:
```diff
@@ -121,6 +121,21 @@ func (m *MPI) EnforceMLPolicy(info *runtime.Info, trainJob *trainer.TrainJob) er
+			if len(requests) > 0 {
+				trainerPS.SinglePodRequests = requests
+			}
+		}
```
@sksingh2005 Please can you try to see if this option is going to work for you: https://github.com/kubernetes/component-helpers/blob/master/resource/helpers.go#L53 ?

If resourcesPerNode is set in the TrainJob, you can:
1. Recalculate PodRequests by using `PodRequests` helper function
2. Define custom `ContainerFn` that overrides the `node` container requests using value from ResourcesPerNode API.

sksingh2005 commented:
Thanks for the suggestion @andreyvelich  sir
Initially, I implemented an approach using `PodRequests` helper with a custom `RecalculateSinglePodRequestsWithOverride` method in [runtime.go](cci:7://file:///d:/GSOC/trainer/pkg/runtime/runtime.go:0:0-0:0) that stored the `TypedPodSpec` and allowed overriding specific container resources. However, I realized that approach was more complex than needed.
I've now implemented your suggested simpler approach - directly modifying the `jobSetTemplateSpec` in [trainingruntime.go](cci:7://file:///d:/GSOC/trainer/pkg/runtime/core/trainingruntime.go:0:0-0:0) before [WithPodSet()](cci:1://file:///d:/GSOC/trainer/pkg/runtime/runtime.go:134:0-151:1) is called:
```go
// Apply resourcesPerNode from TrainJob to the template spec
if trainJob.Spec.Trainer != nil && trainJob.Spec.Trainer.ResourcesPerNode != nil {
    for j := range jobSetTemplateSpec.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers {
        if jobSetTemplateSpec.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Name == constants.Node {
            jobSetTemplateSpec.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Resources = *trainJob.Spec.Trainer.ResourcesPerNode
            break
        }
    }
}
```
Please review once you get time sir!

sksingh2005 commented on pkg/runtime/framework/plugins/mpi/mpi.go:
```diff
@@ -121,6 +121,21 @@ func (m *MPI) EnforceMLPolicy(info *runtime.Info, trainJob *trainer.TrainJob) er
+		if trainerPS := info.FindPodSetByAncestor(constants.AncestorTrainer); trainerPS != nil {
+			res := trainJob.Spec.Trainer.ResourcesPerNode
+			requests := res.Requests
+			// If requests are empty, we use limits as requests.
```
Sir the unit tests for these plugins (`torch_test.go`, etc.) run the plugins in isolation, bypassing `trainingruntime.go`. Without the plugin code, these tests fail because they don't see the resources set.The tests are Failing. Should  we restore the code in the plugins (`mpi.go`, `torch.go`, `plainml.go`) sir ??

andreyvelich replied:
@sksingh2005 Can you remove changes from plugins please?

tenzen-y commented on pkg/runtime/core/trainingruntime.go:
```diff
@@ -171,6 +171,16 @@ func (r *TrainingRuntime) newRuntimeInfo(
+					if trainJob.Spec.Trainer != nil && trainJob.Spec.Trainer.ResourcesPerNode != nil {
+						for j := range jobSetTemplateSpec.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers {
+							if jobSetTemplateSpec.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Name == constants.Node {
+								jobSetTemplateSpec.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Resources = *trainJob.Spec.Trainer.ResourcesPerNode
```
```suggestion
								jobSetTemplateSpec.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers[j].Resources = *trainJob.Spec.Trainer.ResourcesPerNode.DeepCopy()
```

TrainJob is reused across functions. So, we should clone resources to avoid unexpected modifications.

andreyvelich reviewed (commented):
/lgtm
/approve
